qiskit/transpiler/passes/routing/basic_swap.py

```diff
@@ -64,46 +74,82 @@ def run(self, dag):
-            new_dag.compose(subdag, qubits=order)
-
+            cf_layer = False
+            # handle control flow operations
+            cf_nodes = subdag.op_nodes(op=ControlFlowOp) + subdag.named_nodes("continue_loop")
+            if cf_nodes:
+                for node in cf_nodes:  # should be len()=1
+                    subdag_cf, cf_layout = self._transpile_controlflow_op(
+                        node, current_layout, subdag
+                    )
+                    subdag = subdag_cf
+                    cf_layer = True
+            else:
+                for gate in subdag.two_qubit_ops():
+                    physical_q0 = current_layout[gate.qargs[0]]
+                    physical_q1 = current_layout[gate.qargs[1]]
+                    if self.coupling_map.distance(physical_q0, physical_q1) != 1:
+                        # Insert a new layer with the SWAP(s).
+                        swap_layer = DAGCircuit()
+                        swap_layer.add_qreg(canonical_register)
+
+                        path = self.coupling_map.shortest_undirected_path(physical_q0, physical_q1)
+                        for swap in range(len(path) - 2):
+                            connected_wire_1 = path[swap]
+                            connected_wire_2 = path[swap + 1]
+
+                            qubit_1 = current_layout[connected_wire_1]
+                            qubit_2 = current_layout[connected_wire_2]
+
+                            # create the swap operation
+                            swap_layer.apply_operation_back(
+                                SwapGate(), qargs=[qubit_1, qubit_2], cargs=[]
+                            )
+
+                        # layer insertion
+                        order = current_layout.reorder_bits(new_dag.qubits)
+                        new_dag.compose(swap_layer, qubits=order)
+
+                        # update current_layout
+                        for swap in range(len(path) - 2):
+                            current_layout.swap(path[swap], path[swap + 1])
+            if cf_layer or (subdag.depth() == 1 and subdag.op_nodes()[0].name == "continue_loop"):
+                order = current_layout.reorder_bits(new_dag.qubits)
+                new_dag.compose(subdag, qubits=order)
+                current_layout = cf_layout
+            else:
+                order = current_layout.reorder_bits(new_dag.qubits)
+                new_dag.compose(subdag, qubits=order)
+        self.property_set["final_layout"] = current_layout
         return new_dag
 
+    def _transpile_controlflow_op(self, cf_node, current_layout, subdag):
+        """Initialize recursive pass and transpile based on type of control flow operation.
+        The coupling_map is converted to the virtual qubits of the current layout."""
+        new_coupling = layout_transform(self.coupling_map, current_layout)
+        _pass = self.__class__(new_coupling, initial_layout=None)
```

Comment:
I'm really not a fan of co-opting `CouplingMap` to be over virtual qubits instead of physical qubits (mentioned on the other PR as well).  Coupling maps are specifically of physical qubits - it's explicitly in the docstring of the class, and that gives a proper separation of concerns between mutable `Layout`s and the fixed underlying hardware.

It's difficult for me to reason about the correctness of the logic if the coupling map doesn't represent the actual physical connections, and it feels like it opens lots of opportunities for bugs when converting between the "virtual" qubits of the control-flow routing and the outer-circuit routing.  In this case, can we keep the coupling map fixed, and start the inner routing from the current layout?  That feels much more natural to me, and it uses the classes as intended.